### DISCUSSION TITLE
Headless Server: Reload Server Welcome Message

### DISCUSSION THREAD

**Discussion by dtinth** [jamulussoftware/jamulus#1958] [Ideas]
Jamulus supports setting the welcome message file by specifying a filename to `--welcomemessage`, however it is only read once on server startup, to change the welcome message right now I would have to restart the server, losing the boldface on the list for a day.

I am not sure if this is already possible, but it would be great if the welcome message can be updated on-the-fly on a headless server in some way.

**Comment by ann0see** [comment]
I doubt it’s possible now. Probably requires a larger restructuring of the CLI interface. We could maybe monitor if the file containing the welcome message is changed? 

Maybe someone wants to look at it?

**Reply by dtinth** [reply]
@ann0see I see… I looked at `QFileSystemWatcher` but worry about how reliable it is when some text editors save files by renaming the old file, writing to a new file, and moving the new file in place of the old file. (I am also a C++ noob :joy:)

So I explored a bit of different idea… I saw a proposal to make Jamulus remotely programmable using inter-process communication: https://github.com/jamulussoftware/jamulus/discussions/1789. However I do not prefer to use MQTT and prefer simple JSON-based protocols like JSON-RPC, so I tried experimenting with this a little bit.

Start Jamulus server with `Jamulus -n -s --jsonrpcport 22123`, and then `nc localhost 22123`:

```js
>>> {"method":"jamulusserver/getServerInfo","params":{},"id":1}
<<< {"id":1,"jsonrpc":"2.0","result":{"city":"","country":225,"name":"","registrationStatus":0,"welcomeMessage":""}}

>>> {"method":"jamulusserver/setWelcomeMessage","params":{"welcomeMessage":"meow"},"id":2}
<<< {"id":2,"jsonrpc":"2.0","result":"ok"}

>>> {"method":"jamulusserver/getServerInfo","params":{},"id":3}
<<< {"id":3,"jsonrpc":"2.0","result":{"city":"","country":225,"name":"","registrationStatus":0,"welcomeMessage":"meow"}}
```

(I also plan to programmatically control both the client and server, so `--jsonrpcport` applies to both server and client.)

Here are the current changes I made to Jamulus source code to pull this off: https://github.com/jamulussoftware/jamulus/compare/master...dtinth:json-rpc

If you think this might be a good idea, maybe I can open a PR.

**Reply by ann0see** [reply]
Since the JSON work was already done by you and is expandable, maybe it makes sense to open a PR and wait for a review?
The MQTT idea is probably still valid. Just ping the people or comment on this discussion.

I‘m not that involved in the main project anymore. So the decision is up to other contributors?

**Reply by dtinth** [reply]
@ann0see Thank you for your comment, I will make sure that it works for all my personal use cases first, and then I will open a PR. By the way I am not sure who to ping 😄.

**Reply by ann0see** [reply]
The person who developed the mqtt solution (cdmahoney), Rob-NY who did a similar job somewhere in a linked thread, pljones/softins for the code review. gilgongo/me for documentation (after the code has been reviewed and approved)
